File: C++/1-Arrays&Hashing/Algorithms/3-HashMap-SetOperations/HashMap-SetOperationsPro.cpp
#include <iostream>
#include <vector>
#include <unordered_map>
#include <unordered_set>
#include <algorithm>
#include <string>
#include <random>
#include <chrono>

// 1.  Data Deduplication and Unique Element Identification
//
//      * **Use Case:** Processing large datasets where duplicate entries need to be removed,
//          and unique elements identified.  This is common in data cleaning, log analysis,
//          and database management.
//      * **Implementation:** Using `std::unordered_set` to efficiently store and retrieve
//          unique elements.
//
//      * **Real-World Application:** A system that processes user activity logs.  We want to
//          identify unique user IDs to analyze user engagement without counting duplicates.
std::vector<int> getUniqueUserIds(const std::vector<int>& userIds) {
    std::unordered_set<int> uniqueUserIdsSet;
    std::vector<int> uniqueUserIds;

    for (int userId : userIds) {
        uniqueUserIdsSet.insert(userId); // Inserts only if the element is not already present.
    }

    // Copy the unique IDs from the set to a vector for ordered output (if needed).
    for (int userId : uniqueUserIdsSet) {
        uniqueUserIds.push_back(userId);
    }
    return uniqueUserIds;
}

// 2.  Frequency Counting
//
//      * **Use Case:** Counting the occurrences of each distinct element in a collection.
//          This is fundamental in statistical analysis, text processing, and database indexing.
//      * **Implementation:** Using `std::unordered_map` to store elements as keys and their
//          frequencies as values.
//
//      * **Real-World Application:** A system that analyzes website traffic.  We want to count
//          how many times each unique URL was visited to identify popular pages.
std::unordered_map<std::string, int> countUrlVisits(const std::vector<std::string>& urls) {
    std::unordered_map<std::string, int> urlVisitCounts;

    for (const std::string& url : urls) {
        urlVisitCounts[url]++; //  Increments the count for the URL. If the URL is not
                               //  present, it inserts it with a count of 1.
    }
    return urlVisitCounts;
}

// 3.  Membership Testing
//
//      * **Use Case:** Checking if an element exists within a collection.  This is crucial
//          for validation, search operations, and implementing set operations.
//      * **Implementation:** Using `std::unordered_set` for fast membership checks.
//
//      * **Real-World Application:** A spell checker.  We want to quickly determine if a
//          word is in a dictionary of valid words.
bool isWordValid(const std::string& word, const std::unordered_set<std::string>& dictionary) {
    return dictionary.find(word) != dictionary.end(); // Efficient lookup in the set.
}

// 4.  Caching
//
// * **Use Case**: Implementing a simple cache to store frequently accessed data for fast retrieval.
// * **Implementation**: Using `std::unordered_map` to store key-value pairs.  For simplicity,
//     this example doesn't include eviction policies (like LRU), but focuses on the core
//     caching mechanism.
// * **Real-World Application**: A web server caching frequently accessed files (e.g., images, HTML).
//     This reduces the load on the server and improves response times.
class SimpleCache {
public:
    SimpleCache(size_t capacity) : capacity_(capacity) {}

    std::string get(int key) {
        if (cache_.find(key) != cache_.end()) {
            return cache_[key]; // Return cached value
        }
        return ""; // Return empty string if key not found.  A more robust
                      //  solution might throw an exception or return a special value.
    }

    void put(int key, const std::string& value) {
        if (cache_.size() >= capacity_) {
            // Very basic eviction: Remove a random element.  In a real cache, you'd
            //  use a policy like LRU (Least Recently Used).
            if (!cache_.empty()) {
                auto it = cache_.begin();
                std::advance(it, std::rand() % cache_.size()); // Advance to a random element.
                cache_.erase(it);
            }
        }
        cache_[key] = value; // Insert or update the key-value pair.
    }
private:
    std::unordered_map<int, std::string> cache_;
    size_t capacity_;
};

// 5.  Inverted Index for Text Search
//
//      * **Use Case:** Creating an index that maps words to the documents (or positions)
//          where they appear.  This is the core of search engines.
//      * **Implementation:** Using `std::unordered_map` where keys are words and values are
//          vectors of document IDs.
//
//      * **Real-World Application:** A simplified search engine.  Given a set of documents,
//          we want to quickly find which documents contain a specific word.
std::unordered_map<std::string, std::vector<int>> buildInvertedIndex(const std::vector<std::string>& documents) {
    std::unordered_map<std::string, std::vector<int>> invertedIndex;

    for (int i = 0; i < documents.size(); ++i) {
        std::string document = documents[i];
        //  Basic word splitting (for simplicity).  A real-world application would
        //  use a more sophisticated tokenizer.
        std::string word;
        for (char c : document) {
            if (std::isalpha(c)) {
                word += c;
            } else if (!word.empty()) {
                // Convert to lowercase for case-insensitive search
                std::transform(word.begin(), word.end(), word.begin(), ::tolower);
                invertedIndex[word].push_back(i); // Add document ID to the word's list.
                word.clear();
            }
        }
        if (!word.empty()) { //handle the last word
             std::transform(word.begin(), word.end(), word.begin(), ::tolower);
             invertedIndex[word].push_back(i);
        }
    }
    return invertedIndex;
}

// Function to perform search using the inverted index
std::vector<int> searchDocuments(const std::string& query, const std::unordered_map<std::string, std::vector<int>>& index) {
    std::string lowercaseQuery = query;
    std::transform(lowercaseQuery.begin(), lowercaseQuery.end(), lowercaseQuery.begin(), ::tolower);
    if (index.find(lowercaseQuery) != index.end()) {
        return index.at(lowercaseQuery);
    } else {
        return {}; // Return an empty vector if the word is not found.
    }
}

int main() {
    // Seed for the random number generator.  Good practice to do this once at the start.
    std::srand(static_cast<unsigned int>(std::time(nullptr)));

    // 1. Data Deduplication
    std::vector<int> userIds = {1, 5, 2, 5, 3, 1, 4, 2, 6};
    std::vector<int> uniqueUserIds = getUniqueUserIds(userIds);
    std::cout << "Unique User IDs: ";
    for (int userId : uniqueUserIds) {
        std::cout << userId << " ";
    }
    std::cout << std::endl;

    // 2. Frequency Counting
    std::vector<std::string> urls = {
        "https://www.example.com/",
        "https://www.example.com/about",
        "https://www.example.com/",
        "https://www.example.com/contact",
        "https://www.example.com/",
        "https://www.example.com/about"
    };
    std::unordered_map<std::string, int> urlVisitCounts = countUrlVisits(urls);
    std::cout << "URL Visit Counts:\n";
    for (const auto& pair : urlVisitCounts) {
        std::cout << pair.first << ": " << pair.second << std::endl;
    }

    // 3. Membership Testing
    std::unordered_set<std::string> dictionary = {"apple", "banana", "cherry", "date"};
    std::cout << "Is 'banana' a valid word? " << (isWordValid("banana", dictionary) ? "Yes" : "No") << std::endl;
    std::cout << "Is 'grape' a valid word? " << (isWordValid("grape", dictionary) ? "Yes" : "No") << std::endl;

    // 4. Caching
    SimpleCache cache(3); // Cache with a capacity of 3
    cache.put(1, "value1");
    cache.put(2, "value2");
    cache.put(3, "value3");
    std::cout << "Cache get(1): " << cache.get(1) << std::endl; // Output: value1
    std::cout << "Cache get(4): " << cache.get(4) << std::endl; // Output: (empty string)
    cache.put(4, "value4"); // This will evict one of the existing items (randomly)
    std::cout << "Cache get(2): " << cache.get(2) << std::endl; //May or may not exist

    // 5. Inverted Index for Text Search
    std::vector<std::string> documents = {
        "The quick brown fox jumps over the lazy dog.",
        "A journey of a thousand miles begins with a single step.",
        "The best way to predict the future is to invent it.",
        "An apple a day keeps the doctor away.",
        "To be or not to be, that is the question."
    };
    std::unordered_map<std::string, std::vector<int>> invertedIndex = buildInvertedIndex(documents);

    std::string searchQuery = "the";
    std::vector<int> searchResults = searchDocuments(searchQuery, invertedIndex);
    std::cout << "Documents containing '" << searchQuery << "': ";
    if (searchResults.empty())
    {
        std::cout<< "None";
    }
    for (int docId : searchResults) {
        std::cout << docId << " ";
    }
    std::cout << std::endl;

    searchQuery = "future";
    searchResults = searchDocuments(searchQuery, invertedIndex);
    std::cout << "Documents containing '" << searchQuery << "': ";
    if (searchResults.empty())
    {
        std::cout<< "None";
    }
    for (int docId : searchResults) {
        std::cout << docId << " ";
    }
    std::cout << std::endl;

    return 0;
}

File: Java/1-Arrays&Hashing/Algorithms/3-HashMap-SetOperations/HashMap-SetOperationsPro.java
import java.util.*;

public class HashMapSetOperations {

    public static void main(String[] args) {
        // 1. Basic HashMap Operations: Inventory Management
        System.out.println("1. Basic HashMap Operations: Inventory Management");
        inventoryManagement();

        // 2. HashMap for Counting: Word Frequency Analysis
        System.out.println("\n2. HashMap for Counting: Word Frequency Analysis");
        wordFrequencyAnalysis("The quick brown fox jumps over the lazy fox.");

        // 3. HashSet for Membership Test: User Role Management
        System.out.println("\n3. HashSet for Membership Test: User Role Management");
        userRoleManagement();

        // 4. HashMap for Caching: Simple Cache Implementation
        System.out.println("\n4. HashMap for Caching: Simple Cache Implementation");
        simpleCacheImplementation();

        // 5. Using HashMap to solve a problem: Two Sum
        System.out.println("\n5. Using HashMap to solve a problem: Two Sum");
        int[] nums = {2, 7, 11, 15};
        int target = 9;
        int[] result = twoSum(nums, target);
        if (result != null) {
            System.out.println("Two Sum indices: [" + result[0] + ", " + result[1] + "]");
        } else {
            System.out.println("No two sum solution found.");
        }
    }

    // 1. Basic HashMap Operations: Inventory Management
    //   - Real-world application: Managing product stock in an e-commerce system.
    //   - Key operations: put (add/update stock), get (retrieve stock), remove (remove product),
    //                    containsKey (check if product exists).
    public static void inventoryManagement() {
        HashMap<String, Integer> inventory = new HashMap<>();

        // Add initial stock
        inventory.put("Laptop", 100);
        inventory.put("Mouse", 200);
        inventory.put("Keyboard", 150);
        System.out.println("Initial Inventory: " + inventory);

        // Update stock
        inventory.put("Laptop", 120); // Increase laptop stock
        System.out.println("Updated Laptop Stock: " + inventory);

        // Retrieve stock
        int mouseStock = inventory.get("Mouse");
        System.out.println("Mouse Stock: " + mouseStock);

        // Check if a product exists
        if (inventory.containsKey("Keyboard")) {
            System.out.println("Keyboard exists in inventory.");
        }

        // Remove a product
        inventory.remove("Mouse");
        System.out.println("Inventory after removing Mouse: " + inventory);

        //Check for non-existent product
        if(!inventory.containsKey("Monitor")){
            System.out.println("Monitor does not exist in inventory");
        }
        // Use getOrDefault
        int monitorStock = inventory.getOrDefault("Monitor", 0);  //returns 0 if "Monitor" key does not exist
        System.out.println("Monitor Stock: " + monitorStock);
    }

    // 2. HashMap for Counting: Word Frequency Analysis
    //   - Real-world application: Analyzing text data (e.g., log files, documents) to count occurrences of words or events.
    //   - Key idea: Use words as keys and their frequencies as values.
    public static void wordFrequencyAnalysis(String text) {
        HashMap<String, Integer> wordFrequencies = new HashMap<>();

        // Split the text into words
        String[] words = text.toLowerCase().split("\\s+"); // Use lowercase for case-insensitivity

        // Count word frequencies
        for (String word : words) {
            // If the word is already in the map, increment its count; otherwise, add it with a count of 1.
            wordFrequencies.put(word, wordFrequencies.getOrDefault(word, 0) + 1);
        }

        // Print word frequencies
        System.out.println("Word Frequencies: " + wordFrequencies);
    }

    // 3. HashSet for Membership Test: User Role Management
    //   - Real-world application: Managing user permissions and roles in a web application.
    //   - Key idea: Use a HashSet to store roles for each user and efficiently check if a user has a specific role.
    public static void userRoleManagement() {
        HashMap<String, HashSet<String>> userRoles = new HashMap<>();

        // Add user roles
        HashSet<String> johnRoles = new HashSet<>();
        johnRoles.add("admin");
        johnRoles.add("editor");
        userRoles.put("john", johnRoles);

        HashSet<String> janeRoles = new HashSet<>();
        janeRoles.add("editor");
        janeRoles.add("viewer");
        userRoles.put("jane", janeRoles);

        // Check user roles
        String userToCheck = "john";
        String roleToCheck = "admin";
        if (userRoles.containsKey(userToCheck) && userRoles.get(userToCheck).contains(roleToCheck)) {
            System.out.println(userToCheck + " has the role: " + roleToCheck);
        } else {
            System.out.println(userToCheck + " does not have the role: " + roleToCheck);
        }

        userToCheck = "jane";
        roleToCheck = "admin";
        if (userRoles.containsKey(userToCheck) && userRoles.get(userToCheck).contains(roleToCheck)) {
            System.out.println(userToCheck + " has the role: " + roleToCheck);
        } else {
            System.out.println(userToCheck + " does not have the role: " + roleToCheck);
        }
    }

    // 4. HashMap for Caching: Simple Cache Implementation
    //   - Real-world application: Implementing a simple cache to store frequently accessed data.
    //   - Key idea: Use a HashMap to store data with keys, providing fast lookups.  A simple example,
    //                more robust caches use techniques like LRU eviction.
    public static void simpleCacheImplementation() {
        HashMap<String, String> cache = new HashMap<>();

        // Add data to cache
        cache.put("user1", "John Doe");
        cache.put("user2", "Jane Smith");
        System.out.println("Cache: " + cache);

        // Retrieve data from cache
        String user1Data = cache.get("user1");
        if (user1Data != null) {
            System.out.println("Retrieved user1 data from cache: " + user1Data);
        } else {
            System.out.println("user1 data not found in cache.  Fetching from database...");
            // Simulate fetching from a database
            String user1DataFromDB = "John Doe"; // Assume this is fetched from DB
            cache.put("user1", user1DataFromDB); // Store in cache
            System.out.println("Fetched and stored user1 data in cache: " + user1DataFromDB);
        }

        // Check for non-existent data
        String user3Data = cache.get("user3");
        if (user3Data != null) {
            System.out.println("Retrieved user3 data from cache: " + user3Data);
        } else {
            System.out.println("user3 data not found in cache.");
        }
    }

    // 5. Using HashMap to solve a problem: Two Sum
    //   - Problem: Given an array of integers, find two numbers that add up to a specific target.
    //   - Solution: Use a HashMap to store each number and its index.  For each number, check if
    //               the complement (target - number) exists in the HashMap.
    public static int[] twoSum(int[] nums, int target) {
        HashMap<Integer, Integer> numMap = new HashMap<>();

        for (int i = 0; i < nums.length; i++) {
            int complement = target - nums[i];
            if (numMap.containsKey(complement)) {
                return new int[]{numMap.get(complement), i};
            }
            numMap.put(nums[i], i);
        }
        return null; // No solution found
    }
}

